test/unittest/tests/block/test_block_session.cpp

```diff
@@ -0,0 +1,505 @@
+ *
+ * The block session manages an internal caching mechanism for both block and size blocks that are
+ * created or discarded.
+ */
+#include "wt_internal.h"
+#include <catch2/catch.hpp>
+#include "../wrappers/mock_session.h"
+
+void
+validate_and_cleanup_ext_block(WT_EXT *ext)
+{
+    REQUIRE(ext != nullptr);
+    REQUIRE(ext->depth != 0);
+    __wt_free(nullptr, ext);
+}
+
+void
+validate_and_cleanup_size_block(WT_SIZE *size)
+{
+    REQUIRE(size != nullptr);
+    __wt_free(nullptr, size);
+}
+
+void
+cleanup_ext_list(WT_BLOCK_MGR_SESSION *bms)
+{
+    WT_EXT *curr = bms->ext_cache;
+    WT_EXT *tmp;
+    while (curr != NULL) {
+        tmp = curr;
+        curr = curr->next[0];
+        __wt_free(nullptr, tmp);
+    }
+}
+
+void
+validate_and_cleanup_ext_list(WT_BLOCK_MGR_SESSION *bms, int expected_items)
+{
+    int i = 0;
+
+    REQUIRE(bms != nullptr);
+    WT_EXT *curr = bms->ext_cache;
+
+    REQUIRE(bms->ext_cache_cnt == expected_items);
+    for (; i < expected_items; i++) {
+        REQUIRE(bms->ext_cache != nullptr);
+        curr = curr->next[0];
+    }
+    REQUIRE(curr == nullptr);
+    cleanup_ext_list(bms);
+}
+
+void
+cleanup_size_list(WT_BLOCK_MGR_SESSION *bms)
+{
+    WT_SIZE *curr = bms->sz_cache;
+    WT_SIZE *tmp;
+    while (curr != NULL) {
+        tmp = curr;
+        curr = curr->next[0];
+        __wt_free(nullptr, tmp);
+    }
+}
+
+void
+validate_and_cleanup_size_list(WT_BLOCK_MGR_SESSION *bms, int expected_items)
+{
+    int i = 0;
+
+    REQUIRE(bms != nullptr);
+    WT_SIZE *curr = bms->sz_cache;
+
+    REQUIRE(bms->sz_cache_cnt == expected_items);
+    for (; i < expected_items; i++) {
+        REQUIRE(bms->sz_cache != nullptr);
+        curr = curr->next[0];
+    }
+    REQUIRE(curr == nullptr);
+    cleanup_size_list(bms);
+}
+
+TEST_CASE("Block session: __block_ext_alloc", "[block_session]")
+{
+    std::shared_ptr<MockSession> session = MockSession::buildTestMockSession();
+    WT_EXT *ext;
+
+    ext = nullptr;
+    __wt_random_init(&session->getWtSessionImpl()->rnd);
+
+    REQUIRE(__ut_block_ext_alloc(session->getWtSessionImpl(), &ext) == 0);
+    validate_and_cleanup_ext_block(ext);
+}
+
+TEST_CASE("Block session: __block_ext_prealloc", "[block_session]")
+{
+    std::shared_ptr<MockSession> session = MockSession::buildTestMockSession();
+    WT_BLOCK_MGR_SESSION *bms = session->setupBlockManagerSession();
+
+    SECTION("Allocate zero extent blocks")
+    {
+        REQUIRE(__ut_block_ext_prealloc(session->getWtSessionImpl(), 0) == 0);
+        validate_and_cleanup_ext_list(bms, 0);
+    }
+
+    SECTION("Allocate one ext blocks")
+    {
+        REQUIRE(__ut_block_ext_prealloc(session->getWtSessionImpl(), 1) == 0);
+        validate_and_cleanup_ext_list(bms, 1);
+    }
+
+    SECTION("Allocate multiple ext blocks")
+    {
+        REQUIRE(__ut_block_ext_prealloc(session->getWtSessionImpl(), 3) == 0);
+        validate_and_cleanup_ext_list(bms, 3);
+    }
+}
```

Comment:
In this test case we don't actually test what happens if you gave it zero after having existing items in the list, it should validate that the list doesn't grow unexpectedly. Similarly we've only gave it increasing allocations we should check that a decreasing value still allocates the correct number of blocks. Imagine for some reason it saves the max number it's been provided and only every allocates that many.

Reply:
Makes sense, thank you for those test cases, I have added them now.